File: python/tests/test_eit_star.py
"""EIT* — reaches the goal on open space, fails when walled off, near-optimal.

Exercises the effort-informed batch planner in isolation (imported directly from
its module so the suite does not depend on the package re-export).
"""

from __future__ import annotations

from pathlib import Path

import pytest
import yaml
from conftest import CONFIG_DIR, grid_from, open_grid

from navigation.core.params import ParamError, ParamSet
from navigation.global_planning.sampling.eit_star import EITStar
from navigation.maps.occupancy_grid import OccupancyGrid2D


def _eit_config(tmp_path: Path, **overrides: int) -> ParamSet:
    """Real eit_star config with reduced batch budgets so the suite stays fast."""
    doc = yaml.safe_load((CONFIG_DIR / "eit_star.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / "eit_star.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _motion_valid_path(grid: OccupancyGrid2D, path: list[tuple[float, float]]) -> bool:
    return all(grid.is_motion_valid(a, b) for a, b in zip(path, path[1:], strict=False))


def test_reaches_goal_on_open_grid(tmp_path: Path) -> None:
    grid = open_grid(10, 10, seed=7)
    start, goal = (0.5, 0.5), (9.0, 9.0)
    params = _eit_config(tmp_path, batch_size=150, max_batches=4)
    result = EITStar(params).plan(grid, start, goal)
    assert result.success
    # start/goal are permanent sample nodes, so the path pins to them exactly.
    assert result.path[0] == pytest.approx(start)
    assert result.path[-1] == pytest.approx(goal)
    assert _motion_valid_path(grid, result.path)
    assert result.cost > 0.0


def test_no_path_when_walled_off(tmp_path: Path) -> None:
    # Fully occupied middle column: no motion can cross it.
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."], seed=3)
    params = _eit_config(tmp_path, batch_size=80, max_batches=3)
    result = EITStar(params).plan(grid, (0.5, 0.5), (4.5, 4.5))
    assert not result.success
    assert result.path == []
    assert result.cost == 0.0


def test_near_straight_line_on_open_grid(tmp_path: Path) -> None:
    # As an asymptotically optimal batch planner, EIT* should return a cost close
    # to the straight-line lower bound on obstacle-free space.
    grid = open_grid(12, 12, seed=5)
    start, goal = (0.5, 0.5), (11.0, 11.0)
    lower_bound = grid.distance(start, goal)
    result = EITStar(_eit_config(tmp_path, batch_size=200, max_batches=4)).plan(grid, start, goal)
    assert result.success
    assert result.cost <= lower_bound * 1.15


def test_bad_default_param_throws(tmp_path: Path) -> None:
    # step_size default below its declared min must fail at load time.
    bad = _write_bad_step_size(tmp_path)
    with pytest.raises(ParamError):
        ParamSet.from_yaml(bad)


def _write_bad_step_size(tmp_path: Path) -> Path:
    doc = yaml.safe_load((CONFIG_DIR / "eit_star.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "step_size":
            entry["default"] = 0.0  # below declared min 0.01
    out = tmp_path / "eit_star_bad.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return out

File: cpp/tests/test_eit_star.cpp
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/sampling/eit_star.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Point;

namespace {

// Open n x n grid (resolution 0.5 -> n/2 m square); '.' everywhere.
std::vector<std::string> open_rows(int n = 10) {
  return std::vector<std::string>(static_cast<size_t>(n), std::string(static_cast<size_t>(n), '.'));
}

// Reduced batch budget keeps the anytime planner fast while still exercising real
// behavior (informed batches, dual reverse Dijkstra, lexicographic forward search).
std::string eit_cfg() {
  return test::write_temp(
      "eit_star.yaml",
      "algorithm: eit_star\ncategory: global_planning\nparams:\n"
      "  - {name: batch_size, type: int, default: 120, min: 1, max: 100000, description: b}\n"
      "  - {name: max_batches, type: int, default: 5, min: 1, max: 10000, description: mb}\n"
      "  - {name: gamma, type: float, default: 30.0, min: 0.01, max: 1000, description: g}\n"
      "  - {name: step_size, type: float, default: 0.5, min: 0.01, max: 100, description: ss}\n"
      "  - {name: seed, type: int, default: 7, min: 0, max: 2147483647, description: s}\n");
}

}  // namespace

// (a) valid, near-optimal path on open space ----------------------------------

TEST(EitStar, FindsValidNearOptimalPath) {
  auto g = test::make_grid(open_rows());
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::EitStarPlanner p(core::ParamSet::from_yaml(eit_cfg()));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  // start/goal are permanent sample nodes, so the path pins to them exactly.
  ASSERT_GE(r.path.size(), 2u);
  EXPECT_NEAR(r.path.front().x, start.x, 1e-9);
  EXPECT_NEAR(r.path.front().y, start.y, 1e-9);
  EXPECT_NEAR(r.path.back().x, goal.x, 1e-9);
  EXPECT_NEAR(r.path.back().y, goal.y, 1e-9);
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1])) << "path segment must be collision-free";
  }
  // On an obstacle-free grid an asymptotically optimal batch planner returns close
  // to the straight-line lower bound. 1.3x leaves margin for RGG discretization
  // (samples are not aligned to the straight line) at this modest sample budget.
  double lower_bound = g.distance(start, goal);
  EXPECT_LE(r.cost, lower_bound * 1.3);
}

// (b) no-path case: a full wall separates start from goal ---------------------

TEST(EitStar, NoPathAcrossFullWall) {
  auto rows = open_rows();
  rows[5] = "##########";  // impassable horizontal wall
  auto g = test::make_grid(rows);
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::EitStarPlanner p(core::ParamSet::from_yaml(eit_cfg()));
  auto r = p.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
  EXPECT_EQ(r.cost, 0.0);
}

// (c) param validation failure: out-of-range default throws at load ----------

TEST(EitStar, BadStepSizeThrows) {
  std::string bad = test::write_temp(
      "eit_star.yaml",
      "algorithm: eit_star\ncategory: global_planning\nparams:\n"
      "  - {name: step_size, type: float, default: 0.0, min: 0.01, max: 100, description: below min}\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}
